pkg/service/model_build_listener.go

```diff
@@ -0,0 +1,110 @@
+package service
+
+import (
+	"context"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	elbv2api "sigs.k8s.io/aws-load-balancer-controller/apis/elbv2/v1beta1"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/annotations"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/k8s"
+	elbv2model "sigs.k8s.io/aws-load-balancer-controller/pkg/model/elbv2"
+	"strconv"
+)
+
+func (t *defaultModelBuildTask) buildListeners(ctx context.Context, ec2Subnets []*ec2.Subnet) error {
+	tgAttrs, err := t.buildTargetGroupAttributes(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to build target group attributes")
+	}
+	modelTargetType := t.buildTargetType(ctx)
+	apiTargetType := elbv2api.TargetType(modelTargetType)
+	preserveClientIP, err := t.buildPreserveClientIPFlag(ctx, apiTargetType)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to get preserve Client IP configuration")
+	}
+	var certificateARNs []string
+	t.annotationParser.ParseStringSliceAnnotation(annotations.SvcLBSuffixSSLCertificate, &certificateARNs, t.service.Annotations)
+
+	var sslPorts []string
+	t.annotationParser.ParseStringSliceAnnotation(annotations.SvcLBSuffixSSLPorts, &sslPorts, t.service.Annotations)
+	sslPortsSet := sets.NewString(sslPorts...)
+
+	backendProtocol := ""
+	t.annotationParser.ParseStringAnnotation(annotations.SvcLBSuffixBEProtocol, &backendProtocol, t.service.Annotations)
+
+	targetGroupMap := map[string]*elbv2model.TargetGroup{}
+
+	for _, port := range t.service.Spec.Ports {
+		hc, err := t.buildTargetHealthCheck(ctx)
+		tgProtocol := elbv2model.Protocol(port.Protocol)
+		listenerProtocol := elbv2model.Protocol(port.Protocol)
+		if err != nil {
+			return err
+		}
+
+		if tgProtocol != elbv2model.ProtocolUDP && certificateARNs != nil && (sslPortsSet.Len() == 0 || sslPortsSet.Has(port.Name) || sslPortsSet.Has(strconv.Itoa(int(port.Port)))) {
+			if backendProtocol == "ssl" {
+				tgProtocol = elbv2model.ProtocolTLS
+			}
+			listenerProtocol = elbv2model.ProtocolTLS
+		}
+
+		svcPort := intstr.FromInt(int(port.Port))
+		tgName := t.buildTargetGroupName(ctx, port.TargetPort, modelTargetType, tgProtocol, hc)
+		tgResId := t.buildTargetGroupResourceID(k8s.NamespacedName(t.service), svcPort)
```

Comment:
1.  let's move the target group related logic into model_build_target_group.go.
2. let's use tgResourceID to check whether we need to build a new target group or not. (we can save built targetGroup by resourceID), similar to what we have in Ingresses.
3. doing above with change the behavior in cases like two different service port refers to same targetPort. with this existing code, only one targetGroup will be created. after the change, two targetGroup will be created. However, i think the later case should be desired, as we might have annotations to set different tgAttributes per svc port.

Reply:
I had this refactoring as a separate change. I will merge with this PR.

Reply:
tgResourceID for service is namespace/name:svcPort, so there won't be duplicates. As per our offline discussion, we are choosing to create an individual target group per service port. This will allow us to have different configurations for each of the target groups in future.